src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs

```diff
@@ -191,52 +195,33 @@ internal class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
-                    {
-                        _logger.LogCritical("Stream size for resource of type: {ResourceTypeName} is less than 50 bytes, request method: {RequestMethod}", resource.ResourceTypeName, resource.Request.Method);
-                        throw new ServiceUnavailableException();
-                    }
-
-                    PopulateUpsertResourceCommand(sqlCommandWrapper, resource, keepHistory, existingVersion, stream, _coreFeatures.SupportsResourceChangeCapture);
-
                     try
                     {
-                        var newVersion = (int?)await sqlCommandWrapper.ExecuteScalarAsync(cancellationToken);
-                        if (newVersion == null)
+                        if (_schemaInformation.Current >= SchemaVersionConstants.Merge)
                         {
-                            // indicates a redundant delete
-                            return null;
-                        }
-
-                        if (newVersion == -1)
-                        {
-                            // indicates that resource content is same - no new version was created
-                            // We need to send the existing resource in the response matching the correct versionId and lastUpdated as the one stored in DB
-                            return new UpsertOutcome(existingResource, SaveOutcomeType.Updated);
-                        }
-
-                        resource.Version = newVersion.ToString();
-
-                        SaveOutcomeType saveOutcomeType;
-                        if (newVersion == 1)
-                        {
-                            saveOutcomeType = SaveOutcomeType.Created;
+                            var surrId = (await MergeResourcesBeginTransactionAsync(1, cancellationToken)).MinResourceSurrogateId;
+
+                            resource.LastModified = new DateTimeOffset(ResourceSurrogateIdHelper.ResourceSurrogateIdToLastUpdated(surrId), TimeSpan.Zero);
+                            ReplaceVersionIdAndLastUpdatedInMeta(resource);
+
+                            VLatest.MergeResources.PopulateCommand(
+                                sqlCommandWrapper,
+                                AffectedRows: 0,
+                                RaiseExceptionOnConflict: true,
+                                IsResourceChangeCaptureEnabled: _coreFeatures.SupportsResourceChangeCapture,
+                                tableValuedParameters: _mergeResourcesTvpGeneratorVLatest.Generate(new List<MergeResourceWrapper> { new MergeResourceWrapper(resource, surrId, keepHistory, true) }));
+                            await sqlCommandWrapper.ExecuteNonQueryAsync(cancellationToken);
                         }
                         else
                         {
-                            saveOutcomeType = SaveOutcomeType.Updated;
-                            resource.RawResource.IsMetaSet = false;
```

Comment:
>resource.RawResource.IsMetaSet = false; [](http://example.com/codeflow?start=28&length=39)

This line requires calculation of Meta for updated resource on each resource read. Meta computation is 4x more expensive than unzip. This causes huge performance degradation.